File: apps/go/internal/chapter13/boston.go
package chapter13

import (
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter09"
)

// BostonCategory はカテゴリ値の列。
const BostonCategory = "CRIME"

// StandardizeBoston は CRIME をダミー変数にし、欠損値を列の平均値で補完してから、すべての列を標準化する。
// 正解ラベルを予測しないので、PRICE も 1 つの列として扱い、訓練データとテストデータに分けない。
func StandardizeBoston(table chapter02.Table) ([]chapter02.Features, error) {
	crimes := make([]string, 0, len(table.Rows))

	for _, row := range table.Rows {
		crime, err := row.Text(BostonCategory)
		if err != nil {
			return nil, err
		}

		crimes = append(crimes, crime)
	}

	encoded, err := chapter09.Encode(table, BostonCategory, chapter09.Categories(crimes))
	if err != nil {
		return nil, err
	}

	means, err := chapter02.ColumnMeans(encoded.Rows, encoded.Columns)
	if err != nil {
		return nil, err
	}

	filled, err := chapter02.FillMissing(encoded.Rows, encoded.Columns, means)
	if err != nil {
		return nil, err
	}

	standardizer, err := chapter09.Fit(filled)
	if err != nil {
		return nil, err
	}

	return standardizer.Transform(filled)
}

// LoadBoston は CSV を読み込んで前処理する。
func LoadBoston(csvFile string) ([]chapter02.Features, error) {
	table, err := chapter02.LoadTable(csvFile)
	if err != nil {
		return nil, err
	}

	return StandardizeBoston(table)
}

// ToPoints は特徴量のリストを、1 件を 1 行とする数値の並びにする。
func ToPoints(x []chapter02.Features) [][]float64 {
	points := make([][]float64, len(x))
	for i, features := range x {
		points[i] = features.Values
	}

	return points
}

File: apps/ruby/lib/getting_started_ml/chapter13/boston.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 13 章の Boston データの前処理。
  module Chapter13
    module_function

    # CRIME をダミー変数にし、欠損値を列の平均値で補完してから、すべての列を標準化する。
    # 正解の列 PRICE も分析の対象にするので残し、訓練データとテストデータには分けない。
    def standardize_boston(table)
      crimes = table.rows.map { |row| row.text(Chapter09::BOSTON_CATEGORY) }
      encoded = Chapter09.encode(table, Chapter09::BOSTON_CATEGORY, Chapter09.categories(crimes))
      means = Chapter02.column_means(encoded.rows, encoded.columns)
      filled = Chapter02.fill_missing(encoded.rows, encoded.columns, means)

      Chapter09::Standardizer.fit(filled).transform(filled)
    end

    # CSV を読み込んで前処理する。
    def load_boston(csv_file)
      standardize_boston(Chapter02::Table.load(csv_file))
    end

    # 特徴量の並びを、1 件を 1 行とする行列にする。
    def to_matrix(x)
      Numo::DFloat.cast(x.map(&:values))
    end
  end
end
